doc/man/man4/sua.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/sua.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH SUA 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.SH NAME
.B SUA
\- SS7 SCCP User Adaptation Layer (SUA)
.SH SYNOPSIS
.PP
.PP
.B #include <ss7/sccpi.h>
.br
.B #include <ss7/sccpi_ioctl.h>
.br
.B #include <ss7/sua_ioctl.h>
.HP
.BI "int " sua_stream " = open(" device ", " flags );
.HP
.BI "int " sua_xtistr " = t_open(" device ", " flags );
.HP
.BI "int " sua_socket " = socket(" \fRPF_SS7\fI ", " socket_type ", " \fRSS7PROTO_SUA\fI );
.SH DESCRIPTION
.PP
.PP
.B SUA
is an implementation of the SS7 SCCP User Adaptation Layer (SUA) defined in
draft-ietf-sigtran-sua-14.txt.
.B SUA
provides a back haul of the services of the
.BR sccp (4)
over an IP network using Stream Control Transmission Protocol,
.BR sctp (4),
as defined in RFC 2960.
.[
rfc2960
.]
In addition it provides a building block of redundant and distributed SS7
systems. (See
.IR \*(lqNOTICES\*(rq ,
below.)
.PP
.B SUA
is implemented as a
.IR STREAMS (4)
multiplexing driver.
.BR sccp (4)
or
.BR sctp (4)
streams are linked under the
.B SUA
multiplexing driver using the
.IR I_LINK " or " I_PLINK
commands of
.BR streamio (7),
to provide transport or direct SCCP services to
.BR SUA .
Once linked, the streams must be configured using
.BR sua_ioctl (4).
.PP
.B SUA
provides the following stream types that are directly compatible with the
stream types provided by
.BR sccp (4).
To select one of these stream types,
.I device
can be one of the following:
.TP
.I /dev/sua
This device provides the
.B SCCP
protocol service interface using the N-primitives of the Signalling Connection
Control Part Interface described in
.BR sccpi (7).
This interface is suitable for use with other SS7 modules and drivers.
.\"
.\"
.TP
.IR /dev/sua-npi-cl ", " /dev/sua-npi-co
This device provides the
.B SCCP
protocol service interface using the N-primitives of the Network Provider
Interface,
.[
NPI Revision 2.0.0
.]
.BR npi (7),
used for the Signalling Connection Control Part as described in
.BR npi_sccp (7).
The
.BR npi (7)
interface provided on a stream opening this device is suitable for use with
the
.BR sockets (3)
library and provides the
.B SCCP
interface using sockets as described in
.BR sccp_socket (7).
.sp
.I /dev/sua-npi-cl
provides a connectionless
.B N_CLNS
network service provider supporting protocol classes 0 and 1;
.I /dev/sua-npi-co
provides a connection oriented
.B N_CONS
newtork service provider supporting protocol classes 2 and 3.
.\"
.\"
.\"
.TP
.IR /dev/sua-tpi-cl ", " /dev/sua-tpi-co
This device provides the
.B SCCP
protocol service interface using the T-primitives of the Transport Provider
Interface,
.[
tpi revision 2.0.0 draft 2
.]
.BR tpi (7),
used for the Signalling Connection Control Part as described in
.BR tpi_sccp (7).
The
.BR tpi (7)
interface provided on a stream opening this device is a suitable
.I device
for use with
the
.BR xnet (3)
or
.BR xti (3)
library
.BR t_open (3)
call and provides the
.BR SCCP
interface  using XTI
.[
xti revision 1.0
.]
.[
xns99
.]
as described int
.BR xti_sccp (3).
.sp
.I /dev/sua-tpi-cl
provides a connectionless
.B T_CLTS
transport service provider supporting protocol classes 0 and 1;
.I /dev/sua-tpi-co
provides a connection oriented
.B T_COTS
transport service provider supporting protocol classes 2 and 3.
.\"
.\"
.\"
.PP
.B SUA
provides three socket types by use of the
.BR sockets (3)
library
.BR socket (3)
call.
.B SUA
provides sockets in the protocol family
.B PF_SS7
using the protocol number
.BR SS7PROTO_SUA .
The
.I socket_type
argument can be one of the following:
.TP
.B SOCK_DGRAM
Provides a connection-less datagram services without regard to order.  This is
an
.B SCCP
protocol class 0 service provider.
This socket uses the underlying
.I /dev/sua-tpi-cl
transport service provider.
.TP
.B SOCK_RDM
Provides a connection-less reliable datagram service without regard to order.
This is an
.B SCCP
protocol class 1 service provider.
This socket uses the underlying
.I /dev/sua-tpi-cl
transport service provider.
.TP
.B SOCK_SEQPACKET
Provides a connection-oriented reliable datagram service with or without
receipt confirmation and flow control.  This is an
.B SCCP
protocol class 2 or 3 service provider.
This socket uses the underlying
.I /dev/sua-tpi-co
transport service provider.
.\"
.PP
The following entries in /etc/sock2path provides the information necessary for
.BR socksys (4)
to use these devices:
.\"
.PP
.TS
tab(:);
llllll.
PF_SS7:SOCK_DGRAM:SS7PROTO_SUA:/dev/sua-tpi-cl:tpi:clts
PF_SS7:SOCK_RDM:SS7PROTO_SUA:/dev/sua-tpi-cl:tpi:clts
PF_SS7:SOCK_SEQPACKET:SS7PROTO_SUA:/dev/sua-tpi-cl:tpi:cots
.TE
.PP
For additional information, see
.BR sock2path (5).
.\"
.\"
.\"
.\"
.\"
.SH NOTICES
.PP
.PP
.B SUA
can be configured on a system as a drop in transparent replacement for
.BR sccp (4).
In such a configuration, users opening streams or sockets as defined in
.BR sccp (4)
will be opening
.B sua
streams and sockets, but will be unaware of the difference.  In
.B OpenSS7
.B sua
is used as the primary mechanism of redundnacy and distribution and,
when properly configured (see
.BR sua (8)),
.B sua
functions transparently to the SCCP-User.
.PP
Also, opening and binding an
.B sua
stream or socket may result in the opening and binding of an
.BR sccp (4)
stream on the local or remote host in direct support of the local SCCP-User.
.SH IOCTLS
.PP
.PP
All interfaces support a set of management IO controls using
.BR ioctl (2s)
system calls on the
.IR sua_stream ", " sua_xtistr " or " sua_socket
file descriptor or socket.
These management controls are documented in
.BR sua_ioctl (4).
.SH DEVICES
.PP
.PP
.IR /dev/sua ,
.IR /dev/sua-npi-cl ,
.IR /dev/sua-npi-co ,
.IR /dev/sua-tpi-cl ,
.IR /dev/sua-tpi-co .
.SH FILES
.PP
.PP
.BR <ss7/sccpi.h> ,
.BR <ss7/sua_ioctl.h> ,
.BR <sys/npi.h> ,
.BR <sys/npi_ss7.h> ,
.BR <sys/npi_mtp.h> ,
.BR <sys/npi_sccp.h> ,
.BR <sys/tpi.h.h> ,
.BR <sys/tpi_ss7.h> ,
.BR <sys/tpi_mtp.h> ,
.BR <sys/tpi_sccp.h> ,
.BR <xti.h> ,
.BR <xti_sccp.h> ,
.BR <sys/socket.h> .
.SH "SEE ALSO"
.PP
.PP
.BR ss7 (8),
.BR sccpi (7),
.BR npi_sccp (7),
.BR tpi_sccp (7),
.BR xti_sccp (3),
.BR sccp_socket (7),
.BR sockets (3),
.BR xnet (3),
.BR sua_ioctl (4).
.SH VERSIONS
.PP
.PP
The
.B SUA
driver is specific to the
.B OpenSS7
stack.
This is Version 2 of the interface.
.\"
.\"
.XX
.[
SUA-RFC
.]
.\"
.[
$LIST$
.]
.\"
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/drivers/sua.c
/*****************************************************************************

 @(#) File: src/drivers/sua.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/drivers/sua.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

#include <sys/os7/compat.h>

#include "sua.h"
#include "sua_data.h"

/*
 *  SUA MULTIPLEXOR
 *  -------------------------------------------------------------------------
 *  This is a multiplexing driver for SUA.  When SUA is opened by its control
 *  stream (typically a ua configuration daemon), it provides a control
 *  channel for configuration and routing.  When opened normally, by a user
 *  process, it provides an OpenSS7 SCCP Provider.  OpenSS7 SCCP Provider
 *  streams are I_LINKed and I_PLINKed under the multiplexor for Signalling
 *  Gateway (SG) Use by the SUA control stream or SUA user stream.  SCCP
 *  Provider streams can be opened for Application Server (AS) use.  Transport
 *  streams (typically SCTP) are I_LINKed or I_PLINKed under the multiplexor
 *  for both SG and AS use by the UA/LM control stream.
 */

#define SUA_DESCRIP	"SUA STREAMS Multiplexing Driver"
#define SUA_EXTRA	"Part of the OpenSS7 SS7 Stack for Linux Fast-STREAMS"
#define SUA_REVISION	"OpenSS7 src/drivers/sua.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define SUA_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define SUA_DEVICE	"Part of the OpenSS7 Stack for Linux Fast-STREAMS."
#define SUA_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define SUA_LICENSE	"GPL"
#define SUA_BANNER	SUA_DESCRIP	"\n" \
			SUA_EXTRA	"\n" \
			SUA_REVISION	"\n" \
			SUA_COPYRIGHT	"\n" \
			SUA_DEVICE	"\n" \
			SUA_CONTACT	"\n"
#define SUA_SPLASH	SUA_DEVICE	" - " \
			SUA_REVISION	"\n"

#ifdef LINUX
MODULE_AUTHOR(SUA_CONTACT);
MODULE_DESCRIPTION(SUA_DESCRIP);
MODULE_SUPPORTED_DEVICE(SUA_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(SUA_LICENSE);
#endif
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-sua");
#endif
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif
#endif				/* LINUX */

#define SUA_DRV_ID	CONFIG_STREAMS_SUA_MODID
#define SUA_DRV_NAME	CONFIG_STREAMS_SUA_NAME
#define SUA_CMAJORS	CONFIG_STREAMS_SUA_NMAJORS
#define SUA_CMAJOR_0	CONFIG_STREAMS_SUA_MAJOR
#define SUA_UNITS	CONFIG_STREAMS_SUA_NMINORS

/*
 *  ================================================================================
 *
 *  STREAMS Definitions
 *
 *  ================================================================================
 */

#define DRV_ID		SUA_DRV_ID
#define DRV_NAME	SUA_DRV_NAME
#define CMAJORS		SUA_CMAJORS
#define CMAJOR_0	SUA_CMAJOR_0
#define UNITS		SUA_UNITS
#ifdef MODULE
#define DRV_BANNER	SUA_BANNER
#else				/* MODULE */
#define DRV_BANNER	SUA_SPLASH
#endif				/* MODULE */

static struct module_info sua_minfo = {
	.mi_idnum = DRV_ID,		/* Module ID number */
	.mi_idname = DRV_NAME,		/* Module name */
	.mi_minpsz = 1,			/* Min packet size accepted *//* XXX */
	.mi_maxpsz = 512,		/* Max packet size accepted *//* XXX */
	.mi_hiwat = 8 * 512,		/* Hi water mark *//* XXX */
	.mi_lowat = 1 * 512,		/* Lo water mark *//* XXX */
};

static struct qinit sua_u_rinit = {
	.qi_putp = ua_rput,		/* Read put (msg from below) */
	.qi_srvp = ua_rsrv,		/* Read queue service */
	.qi_qopen = ua_open,		/* Each open */
	.qi_qclose = ua_close,		/* Last close */
	.qi_qadmin = NULL,		/* Admin (not used) */
	.qi_minfo = &sua_minfo,		/* Information */
	.qi_mstat = NULL,		/* Statistics */
};

static struct qinit sua_u_winit = {
	.qi_putp = ua_wput,		/* Write put (msg from above) */
	.qi_srvp = ua_wsrv,		/* Write queue service */
	.qi_qopen = NULL,		/* Each open */
	.qi_qclose = NULL,		/* Last close */
	.qi_qadmin = NULL,		/* Admin (not used) */
	.qi_minfo = &sua_minfo,		/* Information */
	.qi_mstat = NULL,		/* Statistics */
};

static struct qinit sua_l_rinit = {
	.qi_putp = ua_rput,		/* Read put (msg from below) */
	.qi_srvp = ua_rsrv,		/* Read queue service */
	.qi_qopen = NULL,		/* Each open */
	.qi_qclose = NULL,		/* Last close */
	.qi_qadmin = NULL,		/* Admin (not used) */
	.qi_minfo = &sua_minfo,		/* Information */
	.qi_mstat = NULL,		/* Statistics */
};

static struct qinit sua_l_winit = {
	.qi_putp = ua_wput,		/* Write put (msg from above) */
	.qi_srvp = ua_wsrv,		/* Write queue service */
	.qi_qopen = NULL,		/* Each open */
	.qi_qclose = NULL,		/* Last close */
	.qi_qadmin = NULL,		/* Admin (not used) */
	.qi_minfo = &sua_minfo,		/* Information */
	.qi_mstat = NULL,		/* Statistics */
};

MODULE_STATIC struct streamtab sua_info = {
	.st_rdinit = &sua_u_rinit,	/* Upper read queue */
	.st_wrinit = &sua_u_winit,	/* Upper write queue */
	.st_muxrinit = &sua_l_rinit,	/* Lower read queue */
	.st_muxwinit = &sua_l_winit,	/* Lower write queue */
};

/*
 *  =========================================================================
 *  
 *  OPEN and CLOSE
 *  
 *  =========================================================================
 */

static dp_t *sua_opens_list = NULL;
static lp_t *sua_links_list = NULL;

static struct ua_driver sua_dinfo = {
	SUA_CMAJOR_0,			/* Major device number */
	SUA_CMINOR,			/* Number of minor devices */
	sizeof(sccp_t),			/* Private structure size */
	NULL,				/* Current control queue */
	&sua_lmq_u_ops,			/* LMQ User operations */
	&sua_ss7_u_ops,			/* SS7 User operations */
	&sua_opens_list,		/* Opened structures list */
	&sua_links_list			/* LInked structures list */
};

/*
 *  =========================================================================
 *
 *  Registration and initialization
 *
 *  =========================================================================
 */
#ifdef LINUX
/*
 *  Linux Registration
 *  -------------------------------------------------------------------------
 */

unsigned short modid = DRV_ID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module ID for the SUA driver. (0 for allocation.)");

major_t major = CMAJOR_0;

#ifndef module_param
MODULE_PARM(major, "h");
#else
module_param(major, uint, 0444);
#endif
MODULE_PARM_DESC(major, "Device number for the SUA driver. (0 for allocation.)");

/*
 *  Linux Fast-STREAMS Registration
 *  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 */
STATIC struct cdevsw sua_cdev = {
	.d_name = DRV_NAME,
	.d_str = &suainfo,
	.d_flag = D_MP,
	.d_fop = NULL,
	.d_mode = S_IFCHR,
	.d_kmod = THIS_MODULE,
};

STATIC int
sua_register_strdev(major_t major)
{
	int err;

	if ((err = register_strdev(&sua_cdev, major)) < 0)
		return (err);
	return (0);
}

STATIC int
sua_unregister_strdev(major_t major)
{
	int err;

	if ((err = unregister_strdev(&sua_cdev, major)) < 0)
		return (err);
	return (0);
}

MODULE_STATIC void __exit
suaterminate(void)
{
	int err, mindex;

	for (mindex = CMAJORS - 1; mindex >= 0; mindex--) {
		if (sua_majors[mindex]) {
			if ((err = sua_unregister_strdev(sua_majors[mindex])))
				cmn_err(CE_PANIC, "%s: cannot unregister major %d", DRV_NAME,
					sua_majors[mindex]);
			if (mindex)
				sua_majors[mindex] = 0;
		}
	}
	if ((err = sua_term_caches()))
		cmn_err(CE_WARN, "%s: could not terminate caches", DRV_NAME);
	return;
}

MODULE_STATIC int __init
suainit(void)
{
	int err, mindex = 0;

	cmn_err(CE_NOTE, DRV_BANNER);	/* console splash */
	if ((err = sua_init_caches())) {
		cmn_err(CE_WARN, "%s: could not init caches, err = %d", DRV_NAME, err);
		suaterminate();
		return (err);
	}
	for (mindex = 0; mindex < CMAJORS; mindex++) {
		if ((err = sua_register_strdev(sua_majors[mindex])) < 0) {
			if (mindex) {
				cmn_err(CE_WARN, "%s: could not register major %d", DRV_NAME,
					sua_majors[mindex]);
				continue;
			} else {
				cmn_err(CE_WARN, "%s: could not register driver, err = %d",
					DRV_NAME, err);
				suaterminate();
				return (err);
			}
		}
		if (sua_majors[mindex] == 0)
			sua_majors[mindex] = err;
#if 0
		LIS_DEVFLAGS(sua_majors[mindex]) |= LIS_MODFLG_CLONE;
#endif
		if (major == 0)
			major = sua_majors[0];
	}
	return (0);
}

/*
 *  Linux Kernel Module Initialization
 *  -------------------------------------------------------------------------
 */
module_init(suainit);
module_exit(suaterminate);

#endif				/* LINUX */
